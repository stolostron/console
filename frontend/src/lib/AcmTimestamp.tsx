/* Copyright Contributors to the Open Cluster Management project */

import { css } from '@emotion/css'
import { SimpleTimestamp } from './SimpleTimestamp'
import { PluginContext } from './PluginContext'
import { useContext } from 'react'
import { TimestampProps } from '@openshift-console/dynamic-plugin-sdk'

type AcmTimestampProps = TimestampProps & {
  showIcon?: boolean
}

const AcmTimestamp: React.FC<AcmTimestampProps> = ({
  timestamp,
  simple,
  omitSuffix,
  className = '',
  showIcon = false,
}) => {
  const {
    ocpApi: { Timestamp },
  } = useContext(PluginContext)

  return Timestamp ? (
    <Timestamp
      timestamp={timestamp}
      simple={simple}
      omitSuffix={omitSuffix}
      className={`${className}${
        !showIcon
          ? ` ${css({
              '.co-icon-and-text__icon': {
                display: 'none',
              },
            })}`
          : ''
      }`}
    />
  ) : (
    <SimpleTimestamp timestamp={timestamp} />
  )
}

export default AcmTimestamp
