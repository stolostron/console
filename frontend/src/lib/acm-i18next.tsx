/* Copyright Contributors to the Open Cluster Management project */
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
import { useTranslation as useReactI18NextTranslation, Trans as ReactTrans } from 'react-i18next'

export function useTranslation() {
  return useReactI18NextTranslation(process.env.TRANSLATION_NAMESPACE)
}

export function Trans(props: any) {
  const { t } = useTranslation()
  return (
    <ReactTrans {...props} t={t}>
      {props.children}
    </ReactTrans>
  )
}
