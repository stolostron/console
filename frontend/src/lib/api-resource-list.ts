/* Copyright Contributors to the Open Cluster Management project */

import { getBackendUrl, getRequest } from '../resources'

const versionUrl = '/apiPaths'

export interface APIResourceNames {
  [kind: string]: APIResourceMeta
}

export interface APIResourceMeta {
  [kind: string]: { pluralName: string }
}

export function getApiPaths() {
  const url = getBackendUrl() + versionUrl
  return getRequest<APIResourceNames>(url)
}
