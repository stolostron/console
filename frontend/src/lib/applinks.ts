/* Copyright Contributors to the Open Cluster Management project */

import { getBackendUrl, getRequest } from '../resources'

export const applinksUrl = '/console-links'

export type IAppLinksResult = {
    data: Record<string, [IAppSwitcherData]>
}

export interface IAppSwitcherData {
    name: string
    url: string
    icon: string
}

export function getApplinks() {
    const url = getBackendUrl() + applinksUrl
    return getRequest<IAppLinksResult>(url)
}
