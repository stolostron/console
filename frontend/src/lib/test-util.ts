/* Copyright Contributors to the Open Cluster Management project */

import { act, ByRoleMatcher, ByRoleOptions, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Scope } from 'nock/types'

export const waitTimeout = 5 * 1000

const waitForOptions = { timeout: waitTimeout }

// By Text

export async function waitForText(text: string, multipleAllowed?: boolean) {
  if (multipleAllowed) {
    await waitFor(() => expect(screen.queryAllByText(text).length).toBeGreaterThan(0), waitForOptions)
  } else {
    await waitFor(() => expect(screen.getByText(text)).toBeDefined(), waitForOptions)
  }
}

export async function waitForNotText(text: string) {
  await waitFor(() => expect(screen.queryAllByText(text)).toHaveLength(0), waitForOptions)
}

export async function waitForInputByText(text: string, index?: number) {
  if (index !== undefined) {
    await waitFor(() => expect(screen.getAllByText(text).length).toBeGreaterThan(index), waitForOptions)
    await waitFor(() => expect(screen.getAllByText(text)[index]).not.toBeDisabled(), waitForOptions)
    await waitFor(
      () =>
        expect((screen.getAllByText(text)[index] as HTMLInputElement).getAttribute('aria-disabled')).not.toEqual(
          'true'
        ),
      waitForOptions
    )
  } else {
    await waitFor(() => expect(screen.getByText(text)).toBeDefined(), waitForOptions)
    await waitFor(() => expect(screen.getByText(text)).not.toBeDisabled(), waitForOptions)
    await waitFor(
      () => expect((screen.getByText(text) as HTMLInputElement).getAttribute('aria-disabled')).not.toEqual('true'),
      waitForOptions
    )
  }
}

export async function waitForInputByTitle(title: string, index?: number) {
  if (index !== undefined) {
    await waitFor(() => expect(screen.getAllByTitle(title).length).toBeGreaterThan(index), waitForOptions)
    await waitFor(() => expect(screen.getAllByTitle(title)[index]).not.toBeDisabled(), waitForOptions)
    await waitFor(
      () =>
        expect((screen.getAllByTitle(title)[index] as HTMLInputElement).getAttribute('aria-disabled')).not.toEqual(
          'true'
        ),
      waitForOptions
    )
  } else {
    await waitFor(() => expect(screen.getByTitle(title)).toBeDefined(), waitForOptions)
    await waitFor(() => expect(screen.getByTitle(title)).not.toBeDisabled(), waitForOptions)
    await waitFor(
      () => expect((screen.getByTitle(title) as HTMLInputElement).getAttribute('aria-disabled')).not.toEqual('true'),
      waitForOptions
    )
  }
}

export async function clickByText(text: string, index?: number) {
  await waitForInputByText(text, index)
  if (index !== undefined) {
    userEvent.click(screen.getAllByText(text)[index])
  } else {
    userEvent.click(screen.getByText(text))
  }
}

export async function clickByTitle(title: string, index?: number) {
  await waitForInputByTitle(title, index)
  if (index !== undefined) {
    userEvent.click(screen.getAllByTitle(title)[index])
  } else {
    userEvent.click(screen.getByTitle(title))
  }
}

export async function typeByText(text: string, type: string, index?: number) {
  await waitForInputByText(text, index)
  if (index !== undefined) {
    userEvent.type(screen.getAllByText(text)[index], type)
  } else {
    userEvent.type(screen.getByText(text), type)
  }
}

// By Placeholder text

export async function typeByPlaceholderText(text: string, type: string, index?: number) {
  if (index !== undefined) {
    userEvent.type(screen.getAllByPlaceholderText(text)[index], type)
  } else {
    userEvent.type(screen.getByPlaceholderText(text), type)
  }
}

export async function clickByPlaceholderText(text: string, index?: number) {
  if (index !== undefined) {
    userEvent.click(screen.getAllByPlaceholderText(text)[index])
  } else {
    userEvent.click(screen.getByPlaceholderText(text))
  }
}

// By Role

export async function waitForRole(role: ByRoleMatcher, options?: ByRoleOptions, multipleAllowed?: boolean) {
  if (multipleAllowed) {
    await waitFor(() => expect(screen.queryAllByRole(role, options).length).toBeGreaterThan(0), waitForOptions)
  } else {
    await waitFor(() => expect(screen.getByRole(role, options)).toBeDefined(), waitForOptions)
  }
}

export async function waitForNotRole(role: ByRoleMatcher, options?: ByRoleOptions) {
  await waitFor(() => expect(screen.queryAllByRole(role, options)).toHaveLength(0), waitForOptions)
}

export async function waitForInputByRole(role: ByRoleMatcher, options?: ByRoleOptions, index?: number) {
  if (index !== undefined) {
    await waitFor(() => expect(screen.getAllByRole(role, options).length).toBeGreaterThan(index), waitForOptions)
    await waitFor(() => expect(screen.getAllByRole(role, options)[index]).not.toBeDisabled(), waitForOptions)
    await waitFor(
      () =>
        expect(
          (screen.getAllByRole(role, options)[index] as HTMLInputElement).getAttribute('aria-disabled')
        ).not.toEqual('true'),
      waitForOptions
    )
  } else {
    await waitFor(() => expect(screen.getByRole(role, options)).toBeDefined(), waitForOptions)
    await waitFor(() => expect(screen.getByRole(role, options)).not.toBeDisabled(), waitForOptions)
    await waitFor(
      () =>
        expect((screen.getByRole(role, options) as HTMLInputElement).getAttribute('aria-disabled')).not.toEqual('true'),
      waitForOptions
    )
  }
}

export async function clickByRole(role: ByRoleMatcher, options?: ByRoleOptions, index?: number) {
  await waitForInputByRole(role, options, index)
  if (index !== undefined) {
    userEvent.click(screen.getAllByRole(role, options)[index])
  } else {
    userEvent.click(screen.getByRole(role, options))
  }
}

export async function typeByRole(type: string, role: ByRoleMatcher, options?: ByRoleOptions, index?: number) {
  await waitForInputByRole(role, options, index)
  if (index !== undefined) {
    userEvent.type(screen.getAllByRole(role, options)[index], type)
  } else {
    userEvent.type(screen.getByRole(role, options), type)
  }
}

// By TestId

export async function waitForTestId(text: string, multipleAllowed?: boolean) {
  if (multipleAllowed) {
    await waitFor(() => expect(screen.queryAllByTestId(text).length).toBeGreaterThan(0), waitForOptions)
  } else {
    await waitFor(() => expect(screen.getByTestId(text)).toBeDefined(), waitForOptions)
  }
}

export async function waitForNotTestId(text: string) {
  await waitFor(() => expect(screen.queryAllByTestId(text)).toHaveLength(0), waitForOptions)
}

export async function waitForInputByTestId(text: string, index?: number) {
  if (index !== undefined) {
    await waitFor(() => expect(screen.getAllByTestId(text).length).toBeGreaterThan(index), waitForOptions)
    await waitFor(() => expect(screen.getAllByTestId(text)[index]).not.toBeDisabled(), waitForOptions)
    await waitFor(
      () =>
        expect((screen.getAllByTestId(text)[index] as HTMLInputElement).getAttribute('aria-disabled')).not.toEqual(
          'true'
        ),
      waitForOptions
    )
  } else {
    await waitFor(() => expect(screen.getByTestId(text)).toBeDefined(), waitForOptions)
    await waitFor(() => expect(screen.getByTestId(text)).not.toBeDisabled(), waitForOptions)
    await waitFor(
      () => expect((screen.getByTestId(text) as HTMLInputElement).getAttribute('aria-disabled')).not.toEqual('true'),
      waitForOptions
    )
  }
}

export async function clickByTestId(text: string, index?: number) {
  await waitForInputByTestId(text, index)
  if (index !== undefined) {
    userEvent.click(screen.getAllByTestId(text)[index])
  } else {
    userEvent.click(screen.getByTestId(text))
  }
}

export async function typeByTestId(id: string, type: string, index?: number) {
  await waitForInputByTestId(id, index)
  if (index !== undefined) {
    userEvent.type(screen.getAllByTestId(id)[index], type)
  } else {
    userEvent.type(screen.getByTestId(id), type)
  }
}

export async function clearByTestId(id: string, index?: number) {
  await waitForInputByTestId(id, index)
  if (index !== undefined) {
    userEvent.clear(screen.getAllByTestId(id)[index])
  } else {
    userEvent.clear(screen.getByTestId(id))
  }
}

// By Label Text

export async function waitForLabelText(text: string, multipleAllowed?: boolean) {
  if (multipleAllowed) {
    await waitFor(() => expect(screen.queryAllByLabelText(text).length).toBeGreaterThan(0), waitForOptions)
  } else {
    await waitFor(() => expect(screen.getByLabelText(text)).toBeDefined(), waitForOptions)
  }
}

export async function waitForNotLabelText(text: string) {
  await waitFor(() => expect(screen.queryAllByLabelText(text)).toHaveLength(0), waitForOptions)
}

export async function waitForInputByLabelText(text: string, index?: number) {
  if (index !== undefined) {
    await waitFor(() => expect(screen.getAllByLabelText(text).length).toBeGreaterThan(index), waitForOptions)
    await waitFor(() => expect(screen.getAllByLabelText(text)[index]).not.toBeDisabled(), waitForOptions)
    await waitFor(
      () =>
        expect((screen.getAllByLabelText(text)[index] as HTMLInputElement).getAttribute('aria-disabled')).not.toEqual(
          'true'
        ),
      waitForOptions
    )
  } else {
    await waitFor(() => expect(screen.getByLabelText(text)).toBeDefined(), waitForOptions)
    await waitFor(() => expect(screen.getByLabelText(text)).not.toBeDisabled(), waitForOptions)
    await waitFor(
      () => expect((screen.getByLabelText(text) as HTMLInputElement).getAttribute('aria-disabled')).not.toEqual('true'),
      waitForOptions
    )
  }
}

export async function clickByLabel(text: string, index?: number) {
  await waitForInputByLabelText(text, index)
  if (index !== undefined) {
    userEvent.click(screen.getAllByLabelText(text)[index])
  } else {
    userEvent.click(screen.getByLabelText(text))
  }
}

export async function typeByLabel(text: string, type: string, index?: number) {
  await waitForInputByLabelText(text, index)
  if (index !== undefined) {
    userEvent.type(screen.getAllByLabelText(text)[index], type)
  } else {
    userEvent.type(screen.getByLabelText(text), type)
  }
}

// By Selector
export async function waitForSelector(container: HTMLElement, selector: string) {
  await waitFor(() => expect(container.querySelector(selector)))
}

export async function waitForNoSelector(container: HTMLElement, selector: string) {
  await waitFor(() => expect(container.querySelector(selector)).toHaveLength(0), waitForOptions)
}

export async function waitForValueBySelector(container: HTMLElement, selector: string, value: string | number) {
  await waitFor(() => expect(container.querySelector(selector)).toHaveValue(value))
}

export async function clickBySelector(container: HTMLElement, selector: string) {
  const elem = await waitFor(() => container.querySelector<HTMLButtonElement>(selector))
  elem?.click()
}

// Other

export async function wait(ms = 0) {
  await act(() => {
    return new Promise((resolve) => {
      setTimeout(resolve, ms)
    })
  })
}

export async function waitForCalled(jestMock: jest.Mock) {
  await waitFor(() => expect(jestMock).toHaveBeenCalled(), waitForOptions)
}

// Nocks

export function nocksAreDone(nocks: Scope[]) {
  for (const nock of nocks) {
    if (!nock.isDone()) return false
  }
  return true
}

export async function waitForNocks(nocks: Scope[]) {
  const timeout = waitForOptions.timeout * nocks.length * 3
  const timeoutMsg = (error: Error) => {
    error.message = `!!!!!!!!!!! Test timed out in waitForNocks()--waited ${timeout / 1000} seconds !!!!!!!!!!!!!`
    error.stack = ''
    return error
  }
  await waitFor(() => expect(nocksAreDone(nocks)).toBeTruthy(), { timeout, onTimeout: timeoutMsg })
}

export async function waitForNock(nock: Scope) {
  await waitFor(() => expect(nock.isDone()).toBeTruthy(), waitForOptions)
}

export async function selectAllRows() {
  await clickByRole('checkbox', {}, 0)
}

export async function selectTableRow(row: number) {
  await clickByLabel(`Select row ${row - 1}`)
}

export async function clickBulkAction(text: string) {
  await clickByText('Actions')
  await clickByText(text)
}

export async function clickRowAction(row: number, text: string) {
  await clickByLabel('Actions', row - 1)
  await clickByText(text)
}

export async function selectByText(placeholdText: string, text: string) {
  await clickByPlaceholderText(placeholdText)
  await clickByText(text)
}

export async function clickHostAction(text: string) {
  await clickByText('ai:Add hosts')
  await clickByText(text)
}

export function isCardEnabled(card: HTMLElement) {
  return card.style.cursor === 'pointer'
}
