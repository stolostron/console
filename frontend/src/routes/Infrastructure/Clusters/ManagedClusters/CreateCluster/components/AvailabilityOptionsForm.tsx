/* Copyright Contributors to the Open Cluster Management project */
import { PopoverIcon } from '@openshift-assisted/ui-lib/common'
import { useTranslation } from '../../../../../../lib/acm-i18next'
import { FormGroup, Radio } from '@patternfly/react-core'
import { AcmForm } from '../../../../../../ui-components'
import { css } from '@emotion/css'
import { TFunction } from 'i18next'

const formWrapper = css({
  paddingBottom: '25px',
  '& .pf-c-form__group-label': {
    padding: `0px 0px 6px 0px !important`,
  },
})

type FormControl = {
  active: any
}

type AvailabilityOptionsFormProps = {
  control: FormControl
  handleChange: (control: FormControl) => void
}

const AvailabilityOptionsForm: React.FC<AvailabilityOptionsFormProps> = (props: any) => {
  const { control } = props
  const { t } = useTranslation()

  const handleChange = (_: any, event: any) => {
    const { control, handleChange } = props

    let targetName = ''
    try {
      targetName = event.target.name
    } catch (event) {
      targetName = ''
    }

    if (targetName) {
      if (targetName === 'controllerAvailabilityPolicy') {
        control.active.controller = event.target.value
      } else if (targetName === 'infraAvailabilityPolicy') {
        control.active.infra = event.target.value
      }
    }
    handleChange(control)
  }

  const labelHA = () => {
    return (
      <>
        {t('Highly available')}{' '}
        <PopoverIcon
          bodyContent={t(
            'Highly available means components should be resilient to problems across fault boundaries as defined by the component to which the policy is attached. This usually means running critical workloads with 3 replicas and with little or no toleration of disruption of the component.'
          )}
        />
      </>
    )
  }

  const labelSingle = () => {
    return (
      <>
        {t('Single replica')}{' '}
        <PopoverIcon
          bodyContent={t(
            'Single replica means components are not expected to be resilient to problems across most fault boundaries associated with high availability. This usually means running critical workloads with just 1 replica and with toleration of full disruption of the component.'
          )}
        />
      </>
    )
  }

  return (
    <AcmForm className={formWrapper}>
      <FormGroup role="radiogroup" isInline fieldId="controller" label={t('Controller availability policy')} isRequired>
        <Radio
          id={'controller-ha'}
          data-testid="controller-ha"
          name={'controllerAvailabilityPolicy'}
          label={labelHA()}
          value={'HighlyAvailable'}
          defaultChecked={control.active.controller === 'HighlyAvailable'}
          onChange={handleChange}
        />
        <Radio
          id={'controller-single'}
          data-testid="controller-single"
          name={'controllerAvailabilityPolicy'}
          label={labelSingle()}
          value={'SingleReplica'}
          defaultChecked={control.active.controller === 'SingleReplica'}
          onChange={handleChange}
        />
      </FormGroup>
      <FormGroup
        role="radiogroup"
        isInline
        fieldId="infrastructure"
        label={t('Infrastructure availability policy')}
        isRequired
      >
        <Radio
          id={'infra-ha'}
          data-testid="infra-ha"
          name={'infraAvailabilityPolicy'}
          label={labelHA()}
          value={'HighlyAvailable'}
          defaultChecked={control.active.infra === 'HighlyAvailable'}
          onChange={handleChange}
        />
        <Radio
          data-testid="infra-single"
          id={'infra-single'}
          name={'infraAvailabilityPolicy'}
          label={labelSingle()}
          value={'SingleReplica'}
          defaultChecked={control.active.infra === 'SingleReplica'}
          onChange={handleChange}
        />
      </FormGroup>
    </AcmForm>
  )
}

export default AvailabilityOptionsForm

export const summary = (control: any, t: TFunction) => {
  const { controller, infra } = control.active || {}
  const getDesc = (ctrl: string) => (ctrl === 'HighlyAvailable' ? t('Highly available') : t('Single replica'))

  const controllerDesc = getDesc(controller)
  const infraDesc = getDesc(infra)

  return [
    { term: t('Controller availability policy'), desc: controllerDesc },
    { term: t('Infrastructure availability policy'), desc: infraDesc },
  ]
}
