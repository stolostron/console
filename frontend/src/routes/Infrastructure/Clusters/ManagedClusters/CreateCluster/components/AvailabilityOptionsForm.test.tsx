/* Copyright Contributors to the Open Cluster Management project */
import { render, screen } from '@testing-library/react'
import { RecoilRoot } from 'recoil'
import { MemoryRouter, Route, Routes } from 'react-router-dom-v5-compat'
import { NavigationPath } from '../../../../../../NavigationPath'

import AvailabilityOptionsForm from './AvailabilityOptionsForm'
import { waitForText } from '../../../../../../lib/test-util'
import userEvent from '@testing-library/user-event'

describe('Availability Options Form', () => {
  const handleChange = jest.fn()
  const activeControl = {
    controller: 'HighlyAvailable',
    infra: 'HighlyAvailable',
  }

  const Component = () => {
    return (
      <RecoilRoot>
        <MemoryRouter initialEntries={[NavigationPath.createCluster]}>
          <Routes>
            <Route
              path={NavigationPath.createCluster}
              element={
                <AvailabilityOptionsForm
                  control={{
                    active: activeControl,
                  }}
                  handleChange={handleChange}
                />
              }
            />
          </Routes>
        </MemoryRouter>
      </RecoilRoot>
    )
  }

  test('renders with highly available selected by default', async () => {
    render(<Component />)
    await waitForText('Controller availability policy')

    const controllerRadioHA = screen.getByTestId('controller-ha')
    const infraRadioHA = screen.getByTestId('infra-ha')
    const controllerRadioSingle = screen.getByTestId('controller-single')
    const infraRadioSingle = screen.getByTestId('infra-single')

    expect(controllerRadioHA).toHaveProperty('checked', true)
    expect(infraRadioHA).toHaveProperty('checked', true)

    userEvent.click(controllerRadioSingle)
    userEvent.click(infraRadioSingle)

    expect(controllerRadioSingle).toHaveProperty('checked', true)
    expect(infraRadioSingle).toHaveProperty('checked', true)
  })
})
