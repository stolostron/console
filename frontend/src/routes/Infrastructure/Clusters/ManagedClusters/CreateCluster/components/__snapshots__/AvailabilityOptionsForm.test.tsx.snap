// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Availability Options Form renders 1`] = `
<div>
  <form
    class="pf-c-form css-1i47bvi"
    novalidate=""
  >
    <div
      class="pf-c-form__group"
      role="radiogroup"
    >
      <div
        class="pf-c-form__group-label"
      >
        <label
          class="pf-c-form__label"
          for="controller"
        >
          <span
            class="pf-c-form__label-text"
          >
            Controller availability policy
          </span>
          <span
            aria-hidden="true"
            class="pf-c-form__label-required"
          >
             
            *
          </span>
        </label>
         
      </div>
      <div
        class="pf-c-form__group-control pf-m-inline"
      >
        <div
          class="pf-c-radio"
        >
          <input
            aria-invalid="false"
            checked=""
            class="pf-c-radio__input"
            data-ouia-component-id="OUIA-Generated-Radio-1"
            data-ouia-component-type="PF4/Radio"
            data-ouia-safe="true"
            data-testid="controller-ha"
            id="controller-ha"
            name="controllerAvailabilityPolicy"
            type="radio"
            value="HighlyAvailable"
          />
          <label
            class="pf-c-radio__label"
            for="controller-ha"
          >
            Highly available
             
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                />
              </svg>
            </button>
          </label>
        </div>
        <div
          class="pf-c-radio"
        >
          <input
            aria-invalid="false"
            class="pf-c-radio__input"
            data-ouia-component-id="OUIA-Generated-Radio-2"
            data-ouia-component-type="PF4/Radio"
            data-ouia-safe="true"
            data-testid="controller-single"
            id="controller-single"
            name="controllerAvailabilityPolicy"
            type="radio"
            value="SingleReplica"
          />
          <label
            class="pf-c-radio__label"
            for="controller-single"
          >
            Single replica
             
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
              data-ouia-component-id="OUIA-Generated-Button-plain-2"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                />
              </svg>
            </button>
          </label>
        </div>
      </div>
    </div>
    <div
      class="pf-c-form__group"
      role="radiogroup"
    >
      <div
        class="pf-c-form__group-label"
      >
        <label
          class="pf-c-form__label"
          for="infrastructure"
        >
          <span
            class="pf-c-form__label-text"
          >
            Infrastructure availability policy
          </span>
          <span
            aria-hidden="true"
            class="pf-c-form__label-required"
          >
             
            *
          </span>
        </label>
         
      </div>
      <div
        class="pf-c-form__group-control pf-m-inline"
      >
        <div
          class="pf-c-radio"
        >
          <input
            aria-invalid="false"
            checked=""
            class="pf-c-radio__input"
            data-ouia-component-id="OUIA-Generated-Radio-3"
            data-ouia-component-type="PF4/Radio"
            data-ouia-safe="true"
            data-testid="infra-ha"
            id="infra-ha"
            name="infraAvailabilityPolicy"
            type="radio"
            value="HighlyAvailable"
          />
          <label
            class="pf-c-radio__label"
            for="infra-ha"
          >
            Highly available
             
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
              data-ouia-component-id="OUIA-Generated-Button-plain-3"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                />
              </svg>
            </button>
          </label>
        </div>
        <div
          class="pf-c-radio"
        >
          <input
            aria-invalid="false"
            class="pf-c-radio__input"
            data-ouia-component-id="OUIA-Generated-Radio-4"
            data-ouia-component-type="PF4/Radio"
            data-ouia-safe="true"
            data-testid="infra-single"
            id="infra-single"
            name="infraAvailabilityPolicy"
            type="radio"
            value="SingleReplica"
          />
          <label
            class="pf-c-radio__label"
            for="infra-single"
          >
            Single replica
             
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
              data-ouia-component-id="OUIA-Generated-Button-plain-4"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                />
              </svg>
            </button>
          </label>
        </div>
      </div>
    </div>
  </form>
</div>
`;
